storage/remote/otlptranslator/prometheusremotewrite/helper.go

```diff
@@ -56,182 +57,117 @@ const (
-	for _, label := range labels {
-		finalKey, err := labelNamer.Build(label.Name)
-		if err != nil {
-			return nil, err
-		}
-		if existingValue, alreadyExists := l[finalKey]; alreadyExists {
-			l[finalKey] = existingValue + ";" + label.Value
-		} else {
-			l[finalKey] = label.Value
+
+	if settings.AllowUTF8 {
+		// UTF8 is allowed, so conflicts aren't possible.
+		c.builder.Reset(sortedLabels)
+	} else {
+		// Now that we have sorted and filtered the labels, build the actual list
+		// of labels, and handle conflicts by appending values.
+		c.builder.Reset(labels.EmptyLabels())
+		var sortErr error
+		sortedLabels.Range(func(l labels.Label) {
+			finalKey, err := labelNamer.Build(l.Name)
+			if err != nil && sortErr == nil {
+				sortErr = err
+				return
+			}
+			if existingValue := c.builder.Get(finalKey); existingValue != "" {
+				c.builder.Set(finalKey, existingValue+";"+l.Value)
+			} else {
+				c.builder.Set(finalKey, l.Value)
+			}
+		})
+		if sortErr != nil {
+			return labels.EmptyLabels(), sortErr
 		}
 	}
 
-	for _, lbl := range promotedAttrs {
-		normalized, err := labelNamer.Build(lbl.Name)
-		if err != nil {
-			return nil, err
+	if settings.EnableTypeAndUnitLabels {
+		unitNamer := otlptranslator.UnitNamer{UTF8Allowed: settings.AllowUTF8}
+		if meta.Type != model.MetricTypeUnknown {
+			c.builder.Set("__type__", strings.ToLower(string(meta.Type)))
 		}
-		if _, exists := l[normalized]; !exists {
-			l[normalized] = lbl.Value
+		if meta.Unit != "" {
+			c.builder.Set("__unit__", unitNamer.Build(meta.Unit))
 		}
 	}
+
+	err := settings.PromoteResourceAttributes.addPromotedAttributes(c.builder, resourceAttrs, settings.AllowUTF8)
+	if err != nil {
+		return labels.EmptyLabels(), err
+	}
 	if promoteScope {
-		var rangeErr error
+		var scopeErr error
 		scope.attributes.Range(func(k string, v pcommon.Value) bool {
 			name, err := labelNamer.Build("otel_scope_" + k)
 			if err != nil {
-				rangeErr = err
+				scopeErr = err
 				return false
 			}
-			l[name] = v.AsString()
+			c.builder.Set(name, v.AsString())
 			return true
 		})
-		if rangeErr != nil {
-			return nil, rangeErr
+		if scopeErr != nil {
+			return labels.EmptyLabels(), scopeErr
 		}
 		// Scope Name, Version and Schema URL are added after attributes to ensure they are not overwritten by attributes.
-		l["otel_scope_name"] = scope.name
-		l["otel_scope_version"] = scope.version
-		l["otel_scope_schema_url"] = scope.schemaURL
+		c.builder.Set("otel_scope_name", scope.name)
+		c.builder.Set("otel_scope_version", scope.version)
+		c.builder.Set("otel_scope_schema_url", scope.schemaURL)
 	}
-
-	if settings.EnableTypeAndUnitLabels {
-		unitNamer := otlptranslator.UnitNamer{UTF8Allowed: settings.AllowUTF8}
-		if metadata.Type != prompb.MetricMetadata_UNKNOWN {
-			l["__type__"] = strings.ToLower(metadata.Type.String())
-		}
-		if metadata.Unit != "" {
-			l["__unit__"] = unitNamer.Build(metadata.Unit)
-		}
-	}
-
```

Comment:
Continuation of previous suggestion:
```suggestion

	if settings.EnableTypeAndUnitLabels {
		unitNamer := otlptranslator.UnitNamer{UTF8Allowed: settings.AllowUTF8}
		if meta.Type != model.MetricTypeUnknown {
			c.builder.Set("__type__", strings.ToLower(string(meta.Type)))
		}
		if meta.Unit != "" {
			c.builder.Set("__unit__", unitNamer.Build(meta.Unit))
		}
	}

```